File: compiler/bench/avlinsert.c
// expected: 99994
#include <stdio.h>
#define N 50000

int left_child[N];
int right_child[N];
int ht[N];
int key[N];
int root = -1;
int node_count = 0;

int height(int node) {
    return (node < 0) ? 0 : ht[node];
}

int max(int a, int b) { return (a > b) ? a : b; }

void update_ht(int node) {
    ht[node] = 1 + max(height(left_child[node]), height(right_child[node]));
}

int balance(int node) {
    return height(left_child[node]) - height(right_child[node]);
}

int rotate_right(int y) {
    int x = left_child[y];
    left_child[y] = right_child[x];
    update_ht(y);
    right_child[x] = y;
    update_ht(x);
    return x;
}

int rotate_left(int x) {
    int y = right_child[x];
    right_child[x] = left_child[y];
    update_ht(x);
    left_child[y] = x;
    update_ht(y);
    return y;
}

int rebalance(int node) {
    int bf = balance(node);
    if (bf > 1) {
        if (balance(left_child[node]) < 0) {
            left_child[node] = rotate_left(left_child[node]);
        }
        return rotate_right(node);
    }
    if (bf < -1) {
        if (balance(right_child[node]) > 0) {
            right_child[node] = rotate_right(right_child[node]);
        }
        return rotate_left(node);
    }
    update_ht(node);
    return node;
}

int avl_insert(int val, int node) {
    if (node < 0) {
        key[node_count] = val;
        left_child[node_count] = -1;
        right_child[node_count] = -1;
        ht[node_count] = 1;
        return node_count++;
    }
    if (val < key[node]) {
        left_child[node] = avl_insert(val, left_child[node]);
    } else {
        right_child[node] = avl_insert(val, right_child[node]);
    }
    return rebalance(node);
}

int lcg(int n) {
    return (n * 1103515245 + 12345) & 2147483647;
}

void build_tree() {
    root = -1;
    node_count = 0;
    int seed = 12345;
    for (int i = 0; i < N; i++) {
        seed = lcg(seed);
        root = avl_insert(seed % N, root);
    }
}

long sum_heights() {
    long sum = 0;
    for (int i = 0; i < N; i++) {
        sum += ht[i];
    }
    return sum;
}

int main() {
    build_tree();
    printf("%ld\n", sum_heights());
    return 0;
}

File: compiler/bench/avlinsert.fs
\ expected: 99994
\ AVL insert 50K nodes, sum of final heights

50000 constant N
create left N cells allot
create right N cells allot
create ht N cells allot
create key N cells allot
variable root
variable ncount

: height ( node -- h ) dup 0< if drop 0 else cells ht + @ then ;
: max ( a b -- max ) 2dup < if swap then drop ;
: update-ht ( node -- ) dup cells left + @ height over cells right + @ height max 1+ swap cells ht + ! ;
: balance ( node -- bf ) dup cells left + @ height swap cells right + @ height - ;

: rotate-right ( y -- x )
  dup cells left + @    \ x = y.left
  dup cells right + @   \ x.right
  2 pick cells left + ! \ y.left = x.right
  over update-ht
  swap cells right + !  \ x.right = y
  dup update-ht ;

: rotate-left ( x -- y )
  dup cells right + @   \ y = x.right
  dup cells left + @    \ y.left
  2 pick cells right + !
  over update-ht
  swap cells left + !
  dup update-ht ;

: rebalance ( node -- node' )
  dup balance 1 > if
    dup cells left + @ balance 0< if
      dup cells left + @ rotate-left over cells left + !
    then
    rotate-right exit
  then
  dup balance -1 < if
    dup cells right + @ balance 0> if
      dup cells right + @ rotate-right over cells right + !
    then
    rotate-left exit
  then
  dup update-ht ;

: avl-insert ( val node -- node' )
  dup 0< if
    drop
    ncount @
    over ncount @ cells key + !
    -1 ncount @ cells left + !
    -1 ncount @ cells right + !
    1 ncount @ cells ht + !
    1 ncount +!
    nip exit
  then
  over over cells key + @ < if
    2dup cells left + @ recurse
    swap cells left + !
    nip rebalance
  else
    2dup cells right + @ recurse
    swap cells right + !
    nip rebalance
  then ;

: lcg ( n -- n' ) 1103515245 * 12345 + 2147483647 and ;

: build-tree ( -- )
  -1 root ! 0 ncount !
  12345
  N 0 do
    lcg dup N mod root @ avl-insert root !
  loop
  drop ;

: sum-heights ( -- n )
  0 N 0 do i cells ht + @ + loop ;

: main build-tree sum-heights . cr ;
